File: src/main/java/org/hisp/dhis/rules/models/RuleActionDisplayText.java
package org.hisp.dhis.rules.models;

import com.google.auto.value.AutoValue;

import javax.annotation.Nonnull;
import javax.annotation.CheckForNull;

@AutoValue
public abstract class RuleActionDisplayText
    extends RuleActionText
{

    @Nonnull
    public static RuleActionDisplayText createForFeedback(
        @CheckForNull String content, @CheckForNull String data )
    {
        if ( content == null && data == null )
        {
            throw new IllegalArgumentException( "Both content and data must not be null" );
        }

        return new AutoValue_RuleActionDisplayText( data == null ? "" : data, content == null ? "" : content,
            LOCATION_FEEDBACK_WIDGET );
    }

    @Nonnull
    public static RuleActionDisplayText createForIndicators(
        @CheckForNull String content, @CheckForNull String data )
    {
        if ( content == null && data == null )
        {
            throw new IllegalArgumentException( "Both content and data must not be null" );
        }

        return new AutoValue_RuleActionDisplayText( data == null ? "" : data, content == null ? "" : content,
            LOCATION_INDICATOR_WIDGET );
    }
}

File: src/main/java/org/hisp/dhis/rules/models/RuleActionDisplayText.kt
package org.hisp.dhis.rules.models


data class RuleActionDisplayText(override var content: String?,
                                 override var data: String?,
                                 override var location: String?) : RuleActionText() {

    companion object {

        @JvmStatic
        fun createForFeedback(content: String?, data: String?): RuleActionDisplayText {
            return when {
                content == null && data == null -> throw IllegalArgumentException("Both content and data must not be null")
                else -> RuleActionDisplayText(content ?: "", data ?: "", LOCATION_FEEDBACK_WIDGET)
            }
        }

        @JvmStatic
        fun createForIndicators(content: String?, data: String?): RuleActionDisplayText {
            return when {
                content == null && data == null -> throw IllegalArgumentException("Both content and data must not be null")
                else -> RuleActionDisplayText(content ?: "", data ?: "", LOCATION_INDICATOR_WIDGET)
            }
        }
    }
}